cpp/memilio/mobility/graph.h

```diff
@@ -436,6 +450,131 @@ void print_graph(std::ostream& os, const Graph& g)
+     * 
+     * @param start_node_idx Id of start node
+     * @param end_node_idx Id of end node
+     * @tparam args Additional arguments for edge construction
+     */
+    template <class... Args>
+    void add_edge(size_t start_node_idx, size_t end_node_idx, Args&&... args)
+    {
+        assert(m_nodes.size() > start_node_idx && m_nodes.size() > end_node_idx);
+        m_edges.emplace_back(start_node_idx, end_node_idx, std::forward<Args>(args)...);
+    }
+
+    /**
+     * @brief Build the graph from the added nodes and edges.
+     * 
+     * Sorts the edges and optionally removes duplicate edges (same start and end node indices).
+     * Wihout dupplicate removal, multiple edges between the same nodes are allowed and the order of insertion is stable.
+     * @param make_unique If true, duplicate edges are removed. The last added edge is kept!
+     * @return Graph<NodePropertyT, EdgePropertyT> The constructed graph.
+     */
+    Graph<NodeProperty, EdgeProperty> build(bool make_unique = false)
+    {
+        sort_edges();
+        if (make_unique) {
+            remove_duplicate_edges();
+        }
+        Graph<NodeProperty, EdgeProperty> graph(std::move(m_nodes), std::move(m_edges));
+        return graph;
+    }
+
+private:
+    /**
+     * @brief Sort the edge vector of a graph.
+     * 
+     * Sorts the edges first by start node index, then by end node index. We use stable_sort to keep the order of insertion
+     * for edges with the same start and end node indices.
+     */
+    void sort_edges()
+    {
+        std::stable_sort(m_edges.begin(), m_edges.end(), [](auto&& e1, auto&& e2) {
+            return e1.start_node_idx == e2.start_node_idx ? e1.end_node_idx < e2.end_node_idx
+                                                          : e1.start_node_idx < e2.start_node_idx;
+        });
+    }
+
+    /**
+     * @brief Remove duplicate edges from a sorted edge vector.
+     * 
+     * Copies all the unique edges to a new vector and replaces the original edge vector with it. Unique means that
+     * the start and end node indices are unique. Other edge properties are not checked and may get lost. Only the last 
+     * edge in the vector is kept.
+     */
+    void remove_duplicate_edges()
+    {
+        std::vector<Edge<EdgePropertyT>> unique_edges;
+        unique_edges.reserve(m_edges.size());
+        auto curr_elem = m_edges.begin();
+        auto next_elem = std::adjacent_find(curr_elem, m_edges.end(), [](auto&& e1, auto&& e2) {
+            return e1.start_node_idx == e2.start_node_idx && e1.end_node_idx == e2.end_node_idx;
+        });
+        while (next_elem != m_edges.end()) {
+            std::copy(curr_elem, next_elem, std::back_inserter(unique_edges));
+            curr_elem = next_elem;
+            while ((*curr_elem).start_node_idx == (*next_elem).start_node_idx &&
+                   (*curr_elem).end_node_idx == (*next_elem).end_node_idx && next_elem != m_edges.end()) {
```

Comment:
next_elem != m_edges.end() has to be moved to the front here. Otherwise, this will fail if we have a duplicate in the last element.